File: examples/spi/coremodel-spi.c
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>

#include <coremodel.h>

static void test_spi_cs(void *priv, unsigned csel)
{
}

static int test_spi_xfr(void *priv, unsigned len, uint8_t *wrdata, uint8_t *rddata)
{
    unsigned idx;
    printf("[%d]", len);
    for(idx=0; idx<len; idx++)
        printf(" %02x", wrdata[idx]);
    printf("\n");
    fflush(stdout);
    for(idx=0; idx<len; idx++)
        rddata[idx] = 'A' + (idx & 63);
    return len;
}

static const coremodel_spi_func_t test_spi_func = {
    .cs = test_spi_cs,
    .xfr = test_spi_xfr };
 
int main(int argc, char *argv[])
{
    int res;
    void *handle;

    if(argc != 4) {
        printf("usage: coremodel-spi <address[:port]> <spi> <cs>\n");
        return 1;
    }

    res = coremodel_connect(argv[1]);
    if(res) {
        fprintf(stderr, "error: failed to connect: %s.\n", strerror(-res));
        return 1;
    }

    handle = coremodel_attach_spi(argv[2], atoi(argv[3]), &test_spi_func, NULL, COREMODEL_SPI_BLOCK);
    if(!handle) {
        fprintf(stderr, "error: failed to attach SPI.\n");
        coremodel_disconnect();
        return 1;
    }

    coremodel_mainloop(-1);

    coremodel_detach(handle);
    coremodel_disconnect();

    return 0;
}

File: examples/spi/coremodel-spi.py
from coremodel import *
import time

class Spi(CoreModelSpi):
    def __init__(self, busname, cs, devname):
        super().__init__(busname = busname, cs = cs)
        self.devname = devname

    def cs(self, csel):
        print("cs ", csel)

    def xfr(self, wrdata, rddata):
        for idx, byte in enumerate(wrdata):
            print("WRITE idx [", idx, "] value", hex(byte))

        for idx in range(len(rddata)):
            # maintain c style (uint8_t *) access when writing data to maintain mutability
            rddata[idx] = 0xa0 + (idx & 0x3f)

        return len(rddata)

if __name__ == "__main__":
    if len(sys.argv) != 5:
        print("usage: python3 coremodel-spi.py <address> <port> <spi bus> <chip select>")
        sys.exit(1)

    address =  sys.argv[1]
    port = sys.argv[2]
    busname = sys.argv[3]
    spi_cs = sys.argv[4]

    cm0 = CoreModel("cm0", address, port, "./../../")

    spi0 = Spi(busname, int(spi_cs), "spi-device")

    try:
        cm0.attach(spi0)
    except NameError as e:
        print(str(e))
        sys.exit(1)

    cm0.cycle_time = 100000

    cm0.start()

    for i in range(3):
        time.sleep(1)

    cm0.stop_event.set()
    cm0.join()